File: modified-flood-fill.c
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>
#include <limits.h>
#include "API.h"

int x=0, y=0, turnR=0;
int dx[4] = {0, 1, 0, -1}, dy[4] = {1, 0, -1, 0};
char buffer[50];
int flood[16][16] = {};
char walls[16][16][4];

// A structure to represent a queue
struct Queue {
	int front, rear, size;
	unsigned capacity;
	int* array;
};

// function to create a queue
// of given capacity.
// It initializes size of queue as 0
struct Queue* createQueue(unsigned capacity)
{
	struct Queue* queue = (struct Queue*)malloc(
		sizeof(struct Queue));
	queue->capacity = capacity;
	queue->front = queue->size = 0;

	// This is important, see the enqueue
	queue->rear = capacity - 1;
	queue->array = (int*)malloc(
		queue->capacity * sizeof(int));
	return queue;
}

// Queue is full when size becomes
// equal to the capacity
int isFull(struct Queue* queue)
{
	return (queue->size == queue->capacity);
}

// Queue is empty when size is 0
int isEmpty(struct Queue* queue)
{
	return (queue->size == 0);
}

// Function to add an item to the queue.
// It changes rear and size
void enqueue(struct Queue* queue, int item)
{
	if (isFull(queue))
		return;
	queue->rear = (queue->rear + 1)
				% queue->capacity;
	queue->array[queue->rear] = item;
	queue->size = queue->size + 1;
}

// Function to remove an item from queue.
// It changes front and size
int dequeue(struct Queue* queue)
{
	if (isEmpty(queue))
		return INT_MIN;
	int item = queue->array[queue->front];
	queue->front = (queue->front + 1)
				% queue->capacity;
	queue->size = queue->size - 1;
	return item;
}

// Function to get front of queue
int front(struct Queue* queue)
{
	if (isEmpty(queue))
		return INT_MIN;
	return queue->array[queue->front];
}

// Function to get rear of queue
int rear(struct Queue* queue)
{
	if (isEmpty(queue))
		return INT_MIN;
	return queue->array[queue->rear];
}

//logi print
void logit(char* text) {
    fprintf(stderr, "%s\n", text);
    fflush(stderr);
}

void logWall(int a) {
    fprintf(stderr, "%d. exploration run...\n", a);
    fflush(stderr);
}

//count directions of cell
int countWays(){
    return !API_wallFront() + !API_wallRight() + !API_wallLeft();
}

//move and change position as x, y
void move(){
    API_moveForward();
    x += dx[turnR];
    y += dy[turnR];
}

//left-handed turn
void left_handed(){
    if (!API_wallLeft()) {
        API_turnLeft();
        turnR = (turnR + 3) % 4;
    }
    while (API_wallFront()){ //|| redMap[x+dx[turnR]][y+dy[turnR]]) {
        API_turnRight();
        turnR = (turnR + 1) % 4;
    }
}

//right-handed turn
void right_handed(){
    if(!API_wallRight()){
        API_turnRight;
        turnR = (turnR + 1) % 4;
    }
    while (API_wallFront()){ //|| redMap[x+dx[turnR]][y+dy[turnR]]){
        API_turnLeft();
        turnR = (turnR + 3) % 4;
    }
}

//mark green if explored
void markCell(char color){
    API_setColor(x, y, color);
   /*sprintf(buffer, "%d", countWays());
    API_setText(x, y, buffer);*/
}

//compare cells by distance to 7, 7
int getCost(int turn){
    return abs((x + dx[(turnR + turn) % 4]) - 7) + abs((y + dy[(turnR + turn) % 4]) - 7);
}


//return 1 if entered in finish grid
int isFinish(){
    if((x==7 && y==7) || (x==7 && y==8) || (x==8 && y==7) || (x==8 && y==8)) return 1;
    return 0;
}

int isStart(){
    if(x==0 && y==0) return 1;
    return 0;
}

//flood the grid considering walls
void floodToFinish(){
    int x, y, nx, ny;
    struct Queue* q = createQueue(200);
    for(int i = 0; i < 16; i++)
        for(int j = 0; j < 16; j++)
            flood[i][j] = -1;

    flood[7][7] = 0; flood[7][8] = 0; flood[8][7] = 0; flood[8][8] = 0;
    enqueue(q, 7); enqueue(q, 7); enqueue(q, 7); enqueue(q, 8); enqueue(q, 8); enqueue(q, 7); enqueue(q, 8); enqueue(q, 8);
    
    while(!isEmpty(q)){
        x = dequeue(q); y = dequeue(q);
		sprintf(buffer, "%d", flood[x][y]);
		API_setText(x, y, buffer);
        for(int i = 0; i < 4; i++){
			nx = x + dx[i]; ny = y + dy[i];
			if(0 <= nx && nx < 16 && 0 <= ny && ny < 16 && flood[nx][ny] == -1 && walls[x][y][i] != 1){
				flood[nx][ny] = flood[x][y] + 1;
				enqueue(q, nx); enqueue(q, ny);
			}
		}
    }
}

void floodToStart(){
    int x, y, nx, ny;
    struct Queue* q = createQueue(100);
    for(int i = 0; i < 16; i++)
        for(int j = 0; j < 16; j++)
            flood[i][j] = -1;

    flood[0][0] = 0; enqueue(q, 0); enqueue(q, 0);
    
    while(!isEmpty(q)){
        x = dequeue(q); y = dequeue(q);
		sprintf(buffer, "%d", flood[x][y]);
		API_setText(x, y, buffer);
        for(int i = 0; i < 4; i++){
			nx = x + dx[i]; ny = y + dy[i];
			if(0 <= nx && nx < 16 && 0 <= ny && ny < 16 && flood[nx][ny] == -1 && walls[x][y][i] != 1){
				flood[nx][ny] = flood[x][y] + 1;
				enqueue(q, nx); enqueue(q, ny);
			}
		}
    }
}

void storeWalls(){
    char wx, wy;
    if(API_wallFront()){
        wx = x + dx[turnR]; wy = y + dy[turnR];
        walls[x][y][turnR] = 1;
        if(!(wx & 0xf0 || wy & 0xf0))
            walls[wx][wy][(turnR+2)%4] = 1;
    }
    if(API_wallRight()){
        wx = x + dx[(turnR+1)%4]; wy = y + dy[(turnR+1)%4];
        walls[x][y][(turnR+1)%4] = 1;
        if(!(wx & 0xf0 || wy & 0xf0))
            walls[wx][wy][(turnR+3)%4] = 1;
    }
    if(API_wallLeft()){
        wx = x + dx[(turnR+3)%4]; wy = y + dy[(turnR+3)%4];
        walls[x][y][(turnR+3)%4] = 1;
        if(!(wx & 0xf0 || wy & 0xf0))
            walls[wx][wy][(turnR+1)%4] = 1;
    }
}

int main(int argc, char* argv[]){

    for (int i=0; i<3; i++){
        logWall(i+1);
        while (!isFinish()) {
            storeWalls();
            floodToFinish();
            if(countWays() < 2){
                left_handed();
            }
            else{
                if(flood[x+dx[turnR]][y+dy[turnR]] < flood[x][y] && !API_wallFront());
                else if(flood[x+dx[(turnR+1)%4]][y+dy[(turnR+1)%4]] < flood[x][y] && !API_wallRight()){
                    API_turnRight();
                    turnR = (turnR+1) % 4;
                }
                else if(flood[x+dx[(turnR+3)%4]][y+dy[(turnR+3)%4]] < flood[x][y] && !API_wallLeft()){
                    API_turnLeft();
                    turnR = (turnR+3) % 4;
                }
                else left_handed();
            }
            markCell('G');
            move();
        }
        storeWalls();
        while(!isStart()){
            storeWalls();
            floodToStart();
            if(countWays() < 2){
                left_handed();
            }
            else{
                if(flood[x+dx[turnR]][y+dy[turnR]] < flood[x][y] && !API_wallFront());
                else if(flood[x+dx[(turnR+1)%4]][y+dy[(turnR+1)%4]] < flood[x][y] && !API_wallRight()){
                    API_turnRight();
                    turnR = (turnR+1) % 4;
                }
                else if(flood[x+dx[(turnR+3)%4]][y+dy[(turnR+3)%4]] < flood[x][y] && !API_wallLeft()){
                    API_turnLeft();
                    turnR = (turnR+3) % 4;
                }
                else left_handed();
            }
            markCell('G');
            move();
        }
        storeWalls();
    }

    Sleep(5000);
    logit("Run on optimised path...");
    floodToFinish();
    while(!isFinish()){
        if(countWays() < 2){
            left_handed();
        }
        else{
            if(flood[x+dx[turnR]][y+dy[turnR]] < flood[x][y] && !API_wallFront());
            else if(flood[x+dx[(turnR+1)%4]][y+dy[(turnR+1)%4]] < flood[x][y] && !API_wallRight()){
                API_turnRight();
                turnR = (turnR+1) % 4;
            }
            else if(flood[x+dx[(turnR+3)%4]][y+dy[(turnR+3)%4]] < flood[x][y] && !API_wallLeft()){
                API_turnLeft();
                turnR = (turnR+3) % 4;
            }
            else left_handed();
        }
        markCell('R');
        move();
    }

    return 0;
}

File: modified-flood-fill.py
import API
import sys
import time
#from machine import Pin
#from micropython import const

INF = 0 #INF = const(0)
N = 16 #N = const(16)
dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

#use list as queue

def log(string):
    sys.stderr.write("{}\n".format(string))
    sys.stderr.flush()

#count directions for the current cell
def countWays():
    return (not API.wallFront()) + (not API.wallRight()) + (not API.wallLeft())

#move forward and update position values
def move(x, y, turnR):
    API.moveForward()
    new_x, new_y = x + dx[turnR], y + dy[turnR]
    return new_x, new_y

#left-handed turn
def left_handed(turnR):
    new_turnR = turnR
    if not API.wallLeft():
        API.turnLeft()
        new_turnR = (new_turnR + 3) % 4
    while API.wallFront():
        API.turnRight()
        new_turnR = (new_turnR + 1) % 4
    return new_turnR

#paint the visited cell
def markCell(color, x, y):
    API.setColor(x, y, color)

#check if position == final cells
def isFinish(x, y):
    if (x==7 and y==7) or (x==7 and y==8) or (x==8 and y==7) or (x==8 and y==8):
        return 1
    else:
        return 0

#check if position == start cell
def isStart(x, y):
    if x==0 and y==0:
        return 1
    else:
        return 0

#flood the grid to finish
def floodToFinish(walls):
    flood = [[INF for _ in range(N)] for _ in range(N)]

    flood[7][7], flood[7][8], flood[8][7], flood[8][8] = 1, 1, 1, 1
    q = [(7, 7), (7, 8), (8, 7), (8, 8)]

    while q:
        x, y = q.pop(0)
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < N and 0 <= ny < N and flood[nx][ny] == INF and walls[x][y][i] == 0:
                flood[nx][ny] = flood[x][y] + 1
                q.append((nx, ny))

    return flood

#flood the grid to start
def floodToStart(walls):
    flood = [[INF for _ in range(N)] for _ in range(N)]

    flood[0][0] = 1
    q = [(0, 0)]

    while q:
        x, y = q.pop(0)
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < N and 0 <= ny < N and flood[nx][ny] == INF and walls[x][y][i] == 0:
                flood[nx][ny] = flood[x][y] + 1
                q.append((nx, ny))

    return flood

#store explored wall data
def storeWalls(x, y, turnR, walls):
    new_walls = walls
    if API.wallFront():
        wx, wy = x + dx[turnR], y + dy[turnR]
        new_walls[x][y][turnR] = 1
        if 0 <= wx < N and 0 <= wy < N:
            new_walls[wx][wy][(turnR+2)%4] = 1

    if API.wallRight():
        wx, wy = x + dx[(turnR+1)%4], y + dy[(turnR+1)%4]
        new_walls[x][y][(turnR+1)%4] = 1
        if 0 <= wx < N and 0 <= wy < N:
            new_walls[wx][wy][(turnR+3)%4] = 1

    if API.wallLeft():
        wx, wy = x + dx[(turnR+3)%4], y + dy[(turnR+3)%4]
        new_walls[x][y][(turnR+3)%4] = 1
        if 0 <= wx < N and 0 <= wy < N:
            new_walls[wx][wy][(turnR+1)%4] = 1

    return new_walls

def main():
    x, y, turnR = 0, 0, 0
    walls = [[[0 for _ in range(4)] for _ in range(N)] for _ in range(N)]

    for i in range(3):
        while not isFinish(x, y):
            walls = storeWalls(x, y, turnR, walls)
            if countWays() < 2:
                turnR = left_handed(turnR)
            else:
                flood = floodToFinish(walls)
                if (0 <= x+dx[turnR] < N and 0 <= y+dy[turnR] < N) and flood[x+dx[turnR]][y+dy[turnR]] < flood[x][y] and (not API.wallFront()):
                    pass
                elif (0 <= x+dx[(turnR+1)%4] < N and 0 <= y+dy[(turnR+1)%4] < N) and flood[x+dx[(turnR+1)%4]][y+dy[(turnR+1)%4]] < flood[x][y] and (not API.wallRight()):
                    API.turnRight()
                    turnR = (turnR+1) % 4
                elif (0 <= x+dx[(turnR+3)%4] < N and 0 <= y+dy[(turnR+3)%4] < N) and flood[x+dx[(turnR+3)%4]][y+dy[(turnR+3)%4]] < flood[x][y] and (not API.wallLeft()):
                    API.turnLeft()
                    turnR = (turnR+3) % 4
                else:
                    turnR = left_handed(turnR)
            markCell('G', x, y)
            x, y = move(x, y, turnR)

        while not isStart(x, y):
            walls = storeWalls(x, y, turnR, walls)
            if countWays() < 2:
                turnR = left_handed(turnR)
            else:
                flood = floodToStart(walls)
                if (0 <= x+dx[turnR] < N and 0 <= y+dy[turnR] < N) and flood[x+dx[turnR]][y+dy[turnR]] < flood[x][y] and (not API.wallFront()):
                        pass
                elif (0 <= x+dx[(turnR+1)%4] < N and 0 <= y+dy[(turnR+1)%4] < N) and flood[x+dx[(turnR+1)%4]][y+dy[(turnR+1)%4]] < flood[x][y] and (not API.wallRight()):
                        API.turnRight()
                        turnR = (turnR+1) % 4
                elif (0 <= x+dx[(turnR+3)%4] < N and 0 <= y+dy[(turnR+3)%4] < N) and flood[x+dx[(turnR+3)%4]][y+dy[(turnR+3)%4]] < flood[x][y] and (not API.wallLeft()):
                        API.turnLeft()
                        turnR = (turnR+3) % 4
                else:
                    turnR = left_handed(turnR)
            markCell('G', x, y)
            x, y = move(x, y, turnR)

    time.sleep(2)

    flood = floodToFinish(walls)
    while not isFinish(x, y):
        if countWays() < 2:
            turnR = left_handed(turnR)
        else:
            if (0 <= x+dx[turnR] < N and 0 <= y+dy[turnR] < N) and flood[x+dx[turnR]][y+dy[turnR]] < flood[x][y] and (not API.wallFront()):
                pass
            elif (0 <= x+dx[(turnR+1)%4] < N and 0 <= y+dy[(turnR+1)%4] < N) and flood[x+dx[(turnR+1)%4]][y+dy[(turnR+1)%4]] < flood[x][y] and (not API.wallRight()):
                API.turnRight()
                turnR = (turnR+1) % 4
            elif (0 <= x+dx[(turnR+3)%4] < N and 0 <= y+dy[(turnR+3)%4] < N) and flood[x+dx[(turnR+3)%4]][y+dy[(turnR+3)%4]] < flood[x][y] and (not API.wallLeft()):
                API.turnLeft()
                turnR = (turnR+3) % 4
            else:
                turnR = left_handed(turnR)
        markCell('R', x, y)
        x, y = move(x, y, turnR)

if __name__ == '__main__':
    main()